include/boost/url/impl/url.ipp

```diff
@@ -584,23 +605,29 @@ url::
 set_userinfo(
     string_view s)
 {
-    detail::copied_strings buf(
-        this->string());
-    s = buf.maybe_copy(s);
-    check_invariants();
-    auto const n = pct_encode_bytes(
-        s, detail::userinfo_chars);
-    auto dest = set_userinfo_impl(n);
-    pct_encode(
-        dest,
-        get(id_host).data() - 1,
-        s,
-        detail::userinfo_chars);
-    decoded_[id_user] = s.size();
+    auto sep = s.find_first_of(':');
+    if (sep == string_view::npos)
+    {
+        set_user(s);
+        remove_password();
+    }
+    else
+    {
+        set_user(s.substr(0, sep));
+        set_password(s.substr(sep + 1));
+    }
     check_invariants();
     return *this;
 }
 
+url&
+url::
+set_userinfo(
+    pct_encoded_view s)
+{
+    return set_userinfo(s.encoded());
```

Comment:
First of all shouldn't this be `set_encoded_userinfo`? Because `s.encoded()` can have escapes, and then you will be escaping the escape, which is wrong.

Second, I think that all these places where we are just passing `s.encoded()` are probably incorrect, because there is no guarantee that the CharSet used to percent-encode `s` is the same as the CharSet which is required for the component that we are setting. These functions need to behave as-if the passed encoded string is first decoded, and then re-encoded using the applicable CharSet.

To simplify things before the review, I am ok with removing ALL `set` overloads which accept `pct_encoded_view`, and just opening an issue with a note that we should explore it in the future.

Reply:
> First of all shouldn't this be set_encoded_userinfo?

Replied in the previous [comment](https://github.com/CPPAlliance/url/pull/238#discussion_r927275743)

> these places where we are just passing s.encoded() are probably incorrect

You're basically asking me what I've been asking for a while but everyone has ignored it.

The last comment where I mention this is [here](https://github.com/CPPAlliance/url/pull/238#issuecomment-1192116439)

I have some opinions on what the function should do but we don't often reach the same conclusions, so I preferred to ask.

Reply:
Hmm we will have to sort this 2ma

Reply:
> You're basically asking me what I've been asking for a while but everyone has ignored it.

I didn't see it until I actually had code in front of me